kubernetes-gcp-go/main.go

```diff
@@ -0,0 +1,152 @@
+		gcpRegion, err := providerCfg.Try("region")
+		if err != nil {
+			gcpRegion = "us-central1"
+		}
+		// Get some additional configuration values or use defaults
+		cfg := config.New(ctx, "")
+		nodePoolCount, err := cfg.TryInt("nodePoolCount")
+		if err != nil {
+			nodePoolCount = 1
+		}
+
+		// Create a new network
+		gkeNetwork, err := compute.NewNetwork(ctx, "gke-network", &compute.NetworkArgs{
+			AutoCreateSubnetworks: pulumi.Bool(false),
+			Description:           pulumi.String("A virtual network for your GKE cluster(s)"),
+		})
+		if err != nil {
+			return err
+		}
+
+		// Create a subnet in the network
+		gkeSubnet, err := compute.NewSubnetwork(ctx, "gke-subnet", &compute.SubnetworkArgs{
+			IpCidrRange:           pulumi.String("10.128.0.0/12"),
+			Network:               gkeNetwork.ID(),
+			PrivateIpGoogleAccess: pulumi.Bool(true),
+		})
+		if err != nil {
+			return err
+		}
+
+		// Create a new GKE cluster
+		gkeCluster, err := container.NewCluster(ctx, "gke-cluster", &container.ClusterArgs{
+			AddonsConfig: &container.ClusterAddonsConfigArgs{
+				DnsCacheConfig: &container.ClusterAddonsConfigDnsCacheConfigArgs{
+					Enabled: pulumi.Bool(true),
+				},
+			},
+			BinaryAuthorization: &container.ClusterBinaryAuthorizationArgs{
+				EvaluationMode: pulumi.String("PROJECT_SINGLETON_POLICY_ENFORCE"),
+			},
+			DatapathProvider: pulumi.String("ADVANCED_DATAPATH"),
+			Description:      pulumi.String("A GKE cluster"),
+			InitialNodeCount: pulumi.Int(1),
+			IpAllocationPolicy: &container.ClusterIpAllocationPolicyArgs{
+				ClusterIpv4CidrBlock:  pulumi.String("/14"),
+				ServicesIpv4CidrBlock: pulumi.String("/20"),
+			},
+			Location: pulumi.String(gcpRegion),
+			MasterAuthorizedNetworksConfig: &container.ClusterMasterAuthorizedNetworksConfigArgs{
+				CidrBlocks: container.ClusterMasterAuthorizedNetworksConfigCidrBlockArray{
+					&container.ClusterMasterAuthorizedNetworksConfigCidrBlockArgs{
+						CidrBlock:   pulumi.String("0.0.0.0/0"),
+						DisplayName: pulumi.String("All networks"),
+					},
+				},
+			},
+			Network:        gkeNetwork.Name,
+			NetworkingMode: pulumi.String("VPC_NATIVE"),
+			PrivateClusterConfig: &container.ClusterPrivateClusterConfigArgs{
+				EnablePrivateNodes:    pulumi.Bool(true),
+				EnablePrivateEndpoint: pulumi.Bool(false),
+				MasterIpv4CidrBlock:   pulumi.String("10.100.0.0/28"),
+			},
+			RemoveDefaultNodePool: pulumi.Bool(true),
+			ReleaseChannel: &container.ClusterReleaseChannelArgs{
+				Channel: pulumi.String("STABLE"),
+			},
+			Subnetwork: gkeSubnet.Name,
+			WorkloadIdentityConfig: &container.ClusterWorkloadIdentityConfigArgs{
+				WorkloadPool: pulumi.String(fmt.Sprintf("%v.svc.id.goog", gcpProject)),
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		// Create a GCP Service Account for the node pool
+		gkeNodepoolSa, err := serviceaccount.NewAccount(ctx, "gke-nodepool-sa", &serviceaccount.AccountArgs{
+			AccountId:   pulumi.String("nodepool-1-sa"),
```

Comment:
I'm not sure what the best way around this is, but it looks like this string has to be unique within the scope of a GCP project; I only noticed it because I tried to stand up a TypeScript instance having already created one in YAML. 

It'll probably cause problems for us in test, but that aside, once two people using the same GCP project try to use this template, they'll run into the same issue. I wonder if using the Random provider would help us around it?

Reply:
GCP service accounts do have to be unique that's correct! 

Would it be possible to use the GetAccount method from the Pulumi provider to determine if an account first exists. If it does use the existing account; should the "GetAccount" method return null/nil/undefined/ect then run the create statement? 

Typically gcp recommended best practice would be one service account per process scoped exactly with only the permissions required. Therefore if you take this approach you could add the unique cluster name into to the SA AcountId? Therefore if you made two clusters with different names you would get two cluster SAs with different ids? 

In the real word we try to steal away from uuid/random postfixes and normally recommend customers use a controlled increment.

Reply:
Embedding the cluster name (which _should_ have a unique name due to Pulumi's auto-naming mechanisms) would take care of the issue with the service account, I'd guess.